File: scripts/migrate_bookmarks.py
#!/usr/bin/env python3
"""
浏览器收藏夹迁移脚本
支持 Chrome、Edge、Firefox、Safari、Tabbit、Arc、Brave 等主流浏览器
"""

import json
import shutil
import time
import uuid
import hashlib
import copy
import sqlite3
import plistlib
import argparse
import sys
import os
from pathlib import Path

# macOS 浏览器书签路径
MACOS_BROWSER_PATHS = {
    "chrome": "~/Library/Application Support/Google/Chrome/Default/Bookmarks",
    "edge": "~/Library/Application Support/Microsoft Edge/Default/Bookmarks",
    "firefox": "~/Library/Application Support/Firefox/Profiles",
    "safari": "~/Library/Safari/Bookmarks.plist",
    "tabbit": "~/Library/Application Support/Tabbit/Default/Bookmarks",
    "arc": "~/Library/Application Support/Arc/User Data/Default/Bookmarks",
    "brave": "~/Library/Application Support/BraveSoftware/Brave-Browser/Default/Bookmarks",
}

# Windows 浏览器书签路径
WINDOWS_BROWSER_PATHS = {
    "chrome": "%LOCALAPPDATA%/Google/Chrome/User Data/Default/Bookmarks",
    "edge": "%LOCALAPPDATA%/Microsoft/Edge/User Data/Default/Bookmarks",
    "firefox": "%APPDATA%/Mozilla/Firefox/Profiles",
    "brave": "%LOCALAPPDATA%/BraveSoftware/Brave-Browser/User Data/Default/Bookmarks",
}


def get_bookmarks_path(browser_name, platform=None):
    """获取浏览器书签文件路径"""
    if platform is None:
        platform = sys.platform

    browser_name = browser_name.lower()

    if platform == "darwin":
        paths = MACOS_BROWSER_PATHS
    elif platform.startswith("win"):
        paths = WINDOWS_BROWSER_PATHS
    else:
        raise ValueError(f"Unsupported platform: {platform}")

    if browser_name not in paths:
        raise ValueError(f"Unsupported browser: {browser_name}")

    path = os.path.expanduser(paths[browser_name])

    # Firefox 需要找到 profiles 目录下的 places.sqlite
    if browser_name == "firefox":
        profile_dir = find_firefox_profile(path)
        if profile_dir:
            return os.path.join(profile_dir, "places.sqlite")
        return None

    # Safari 使用 plist
    if browser_name == "safari":
        return path if os.path.exists(path) else None

    return path if os.path.exists(os.path.expanduser(path)) else None


def find_firefox_profile(base_path):
    """查找 Firefox 默认 profile 目录"""
    base_path = os.path.expanduser(base_path)
    if not os.path.exists(base_path):
        return None
    for item in os.listdir(base_path):
        if item.endswith(".default") or item.endswith(".default-release"):
            full_path = os.path.join(base_path, item)
            if os.path.isdir(full_path):
                return full_path
    return None


def load_chromium_bookmarks(path):
    """加载 Chromium 系浏览器书签（JSON格式）"""
    with open(path, 'r', encoding='utf-8') as f:
        return json.load(f)


def load_firefox_bookmarks(path):
    """加载 Firefox 书签（SQLite格式）"""
    conn = sqlite3.connect(path)
    cursor = conn.cursor()

    # 获取所有书签
    cursor.execute("""
        SELECT b.id, b.title, b.parent, b.position, p.url
        FROM moz_bookmarks b
        LEFT JOIN moz_places p ON b.fk = p.id
        WHERE b.type = 1
        ORDER BY b.parent, b.position
    """)

    bookmarks = []
    for row in cursor.fetchall():
        bookmarks.append({
            "id": row[0],
            "name": row[1],
            "parent": row[2],
            "position": row[3],
            "url": row[4]
        })

    # 获取文件夹结构
    cursor.execute("""
        SELECT id, title, parent, position
        FROM moz_bookmarks
        WHERE type = 2
        ORDER BY parent, position
    """)

    folders = {}
    for row in cursor.fetchall():
        folders[row[0]] = {
            "id": row[0],
            "name": row[1],
            "parent": row[2],
            "position": row[3],
            "children": []
        }

    conn.close()
    return {"bookmarks": bookmarks, "folders": folders}


def load_safari_bookmarks(path):
    """加载 Safari 书签（plist格式）"""
    with open(path, 'rb') as f:
        plist = plistlib.load(f)
    return plist


def find_folder(node, name):
    """递归查找指定名称的文件夹"""
    if node.get("type") == "folder" and node.get("name") == name:
        return node
    for child in node.get("children", []):
        result = find_folder(child, name)
        if result:
            return result
    return None


def find_bookmarks_by_keyword(node, keyword):
    """按关键词查找书签"""
    results = []
    keyword_lower = keyword.lower()

    def search(node):
        if node.get("type") == "url":
            name = node.get("name", "").lower()
            url = node.get("url", "").lower()
            if keyword_lower in name or keyword_lower in url:
                results.append(node)
        for child in node.get("children", []):
            search(child)

    search(node)
    return results


def get_max_id(node):
    """获取书签树中的最大ID"""
    max_id = int(node.get("id", 0))
    for child in node.get("children", []):
        max_id = max(max_id, get_max_id(child))
    return max_id


def remap_ids(node, id_counter):
    """重新映射书签ID和GUID"""
    new_node = copy.deepcopy(node)
    new_node["id"] = str(id_counter[0])
    id_counter[0] += 1

    if "guid" in new_node:
        new_node["guid"] = str(uuid.uuid4())

    # 更新时间戳
    now = str(int(time.time() * 1000000) + 116444736000000000)
    if new_node.get("type") == "folder":
        new_node["date_added"] = now
        new_node["date_modified"] = now
    else:
        new_node["date_added"] = now

    # 移除 Chrome 特有的 meta_info
    if "meta_info" in new_node:
        del new_node["meta_info"]

    for i, child in enumerate(new_node.get("children", [])):
        new_node["children"][i] = remap_ids(child, id_counter)

    return new_node


def calculate_checksum(data):
    """计算 Chromium 书签的 checksum"""
    content = json.dumps(data["roots"], ensure_ascii=False, separators=(",", ":"), sort_keys=True)
    return hashlib.md5(content.encode("utf-8")).hexdigest()


def list_all_folders(node, path=""):
    """列出所有文件夹"""
    folders = []
    if node.get("type") == "folder":
        full_path = f"{path}/{node['name']}" if path else node["name"]
        bookmark_count = len([c for c in node.get("children", []) if c.get("type") == "url"])
        folders.append((full_path, bookmark_count))
        for child in node.get("children", []):
            folders.extend(list_all_folders(child, full_path))
    return folders


def migrate_bookmarks(source_browser, target_browser, folder_name=None, keyword=None, dry_run=False):
    """执行书签迁移"""

    # 1. 获取源和目标书签路径
    source_path = get_bookmarks_path(source_browser)
    target_path = get_bookmarks_path(target_browser)

    if not source_path:
        print(f"❌ 源浏览器 '{source_browser}' 书签文件未找到")
        return False

    if not target_path:
        print(f"❌ 目标浏览器 '{target_browser}' 书签文件未找到")
        return False

    print(f"📂 源: {source_path}")
    print(f"📂 目标: {target_path}")

    # 2. 加载源书签
    if source_browser in ["firefox"]:
        source_data = load_firefox_bookmarks(source_path)
        print("⚠️ Firefox 书签迁移需要额外处理，当前版本暂不支持")
        return False
    elif source_browser in ["safari"]:
        source_data = load_safari_bookmarks(source_path)
        print("⚠️ Safari 书签迁移需要额外处理，当前版本暂不支持")
        return False
    else:
        source_data = load_chromium_bookmarks(source_path)

    # 3. 加载目标书签
    if target_browser in ["firefox", "safari"]:
        print("⚠️ 目标浏览器为 Firefox/Safari，暂不支持写入")
        return False
    else:
        target_data = load_chromium_bookmarks(target_path)

    # 4. 查找要迁移的内容
    items_to_migrate = []

    if folder_name:
        # 按文件夹名称查找
        folder = None
        for root_name, root_node in source_data["roots"].items():
            if isinstance(root_node, dict):
                folder = find_folder(root_node, folder_name)
                if folder:
                    break

        if not folder:
            print(f"❌ 未找到文件夹 '{folder_name}'")
            print("\n可用文件夹:")
            for root_name, root_node in source_data["roots"].items():
                if isinstance(root_node, dict):
                    folders = list_all_folders(root_node)
                    for path, count in folders:
                        print(f"  📁 {path} ({count} 个书签)")
            return False

        items_to_migrate = [folder]
        print(f"✅ 找到文件夹 '{folder_name}'，包含 {len(folder.get('children', []))} 个子项")

    elif keyword:
        # 按关键词查找
        for root_name, root_node in source_data["roots"].items():
            if isinstance(root_node, dict):
                results = find_bookmarks_by_keyword(root_node, keyword)
                items_to_migrate.extend(results)

        if not items_to_migrate:
            print(f"❌ 未找到包含关键词 '{keyword}' 的书签")
            return False

        print(f"✅ 找到 {len(items_to_migrate)} 个匹配的书签")

    else:
        # 全量迁移 - 迁移所有根节点下的内容
        for root_name, root_node in source_data["roots"].items():
            if isinstance(root_node, dict) and root_node.get("children"):
                items_to_migrate.extend(root_node["children"])
        print(f"✅ 全量迁移，共 {len(items_to_migrate)} 个顶级项")

    # 5. 预览模式
    if dry_run:
        print("\n🔍 预览模式（不执行写入）:")
        for item in items_to_migrate:
            if item.get("type") == "folder":
                print(f"  📁 {item['name']} ({len(item.get('children', []))} 个子项)")
            else:
                print(f"  🔗 {item.get('name', 'Unnamed')} - {item.get('url', 'No URL')}")
        return True

    # 6. 备份目标书签
    backup_path = target_path + ".bak_" + time.strftime("%Y%m%d_%H%M%S")
    shutil.copy2(target_path, backup_path)
    print(f"\n💾 已备份目标书签到: {backup_path}")

    # 7. 执行迁移
    max_id = 0
    for root_name, root_node in target_data["roots"].items():
        if isinstance(root_node, dict):
            max_id = max(max_id, get_max_id(root_node))

    id_counter = [max_id + 1]

    for item in items_to_migrate:
        new_item = remap_ids(item, id_counter)
        target_data["roots"]["bookmark_bar"]["children"].append(new_item)

    # 8. 更新 checksum 并写入
    target_data["checksum"] = calculate_checksum(target_data)

    with open(target_path, 'w', encoding='utf-8') as f:
        json.dump(target_data, f, ensure_ascii=False, indent=3)

    print(f"✅ 迁移完成！已写入 {len(items_to_migrate)} 个项目到 {target_browser}")
    print(f"📝 请重启 {target_browser} 以刷新书签")
    return True


def main():
    parser = argparse.ArgumentParser(description="浏览器收藏夹迁移工具")
    parser.add_argument("--source", "-s", required=True, help="源浏览器名称 (chrome/edge/tabbit/brave/arc)")
    parser.add_argument("--target", "-t", required=True, help="目标浏览器名称 (chrome/edge/tabbit/brave/arc)")
    parser.add_argument("--folder", "-f", help="指定要迁移的文件夹名称")
    parser.add_argument("--keyword", "-k", help="按关键词过滤书签")
    parser.add_argument("--dry-run", "-d", action="store_true", help="预览模式，不实际写入")
    parser.add_argument("--list", "-l", action="store_true", help="列出源浏览器的所有文件夹")

    args = parser.parse_args()

    if args.list:
        source_path = get_bookmarks_path(args.source)
        if not source_path:
            print(f"❌ 未找到 {args.source} 的书签文件")
            return

        data = load_chromium_bookmarks(source_path)
        print(f"\n📂 {args.source} 的所有收藏夹:")
        for root_name, root_node in data["roots"].items():
            if isinstance(root_node, dict):
                folders = list_all_folders(root_node)
                for path, count in folders:
                    print(f"  📁 {path} ({count} 个书签)")
        return

    migrate_bookmarks(
        source_browser=args.source,
        target_browser=args.target,
        folder_name=args.folder,
        keyword=args.keyword,
        dry_run=args.dry_run
    )


if __name__ == "__main__":
    main()

File: scripts/migrate-bookmarks.mjs
#!/usr/bin/env node
/**
 * 浏览器收藏夹迁移脚本 (Node.js 版本)
 * 支持 Chrome、Edge、Tabbit、Arc、Brave 等 Chromium 系浏览器
 * 也支持 Firefox/Safari 读取
 */

import fs from 'fs';
import path from 'path';
import os from 'os';
import crypto from 'crypto';
import { execSync } from 'child_process';

// ── 浏览器书签路径 ───────────────────────────────────────────
const MACOS_PATHS = {
  chrome:  '~/Library/Application Support/Google/Chrome/Default/Bookmarks',
  edge:    '~/Library/Application Support/Microsoft Edge/Default/Bookmarks',
  tabbit:  '~/Library/Application Support/Tabbit/Default/Bookmarks',
  arc:     '~/Library/Application Support/Arc/User Data/Default/Bookmarks',
  brave:   '~/Library/Application Support/BraveSoftware/Brave-Browser/Default/Bookmarks',
  firefox: '~/Library/Application Support/Firefox/Profiles',
  safari:  '~/Library/Safari/Bookmarks.plist',
};

const WIN_PATHS = {
  chrome:  '%LOCALAPPDATA%/Google/Chrome/User Data/Default/Bookmarks',
  edge:    '%LOCALAPPDATA%/Microsoft/Edge/User Data/Default/Bookmarks',
  brave:   '%LOCALAPPDATA%/BraveSoftware/Brave-Browser/User Data/Default/Bookmarks',
  firefox: '%APPDATA%/Mozilla/Firefox/Profiles',
};

function expandHome(p) {
  return p.replace(/^~/, os.homedir());
}

function getBookmarksPath(browser) {
  const platform = os.platform();
  const name = browser.toLowerCase();
  let raw;

  if (platform === 'darwin') {
    raw = MACOS_PATHS[name];
  } else if (platform.startsWith('win')) {
    raw = WIN_PATHS[name];
  } else {
    console.error(`❌ 不支持的平台: ${platform}`);
    return null;
  }

  if (!raw) {
    console.error(`❌ 不支持的浏览器: ${name}`);
    return null;
  }

  let resolved = expandHome(raw);

  // Firefox 需要找 profile 目录
  if (name === 'firefox') {
    const profileDir = findFirefoxProfile(resolved);
    return profileDir ? path.join(profileDir, 'places.sqlite') : null;
  }

  if (name === 'safari') {
    return fs.existsSync(resolved) ? resolved : null;
  }

  resolved = path.resolve(resolved);
  if (!fs.existsSync(resolved)) {
    console.error(`❌ 书签文件不存在: ${resolved}`);
    console.error('   请确认浏览器已安装并运行过至少一次');
    return null;
  }
  return resolved;
}

function findFirefoxProfile(basePath) {
  basePath = expandHome(basePath);
  if (!fs.existsSync(basePath)) return null;
  for (const item of fs.readdirSync(basePath)) {
    if (item.endsWith('.default') || item.endsWith('.default-release')) {
      const full = path.join(basePath, item);
      if (fs.statSync(full).isDirectory()) return full;
    }
  }
  return null;
}

// ── 加载书签 ────────────────────────────────────────────────
function loadChromiumBookmarks(filePath) {
  return JSON.parse(fs.readFileSync(filePath, 'utf-8'));
}

function loadFirefoxBookmarks(filePath) {
  console.warn('⚠️ Firefox 书签迁移需要额外处理，当前版本暂不支持写入');
  return null;
}

function loadSafariBookmarks(filePath) {
  console.warn('⚠️ Safari 书签迁移需要额外处理，当前版本暂不支持写入');
  return null;
}

// ── 树操作 ──────────────────────────────────────────────────
function findFolder(node, name) {
  if (node.type === 'folder' && node.name === name) return node;
  for (const child of node.children || []) {
    const found = findFolder(child, name);
    if (found) return found;
  }
  return null;
}

function findBookmarksByKeyword(node, keyword) {
  const results = [];
  const kw = keyword.toLowerCase();
  function search(n) {
    if (n.type === 'url') {
      const nm = (n.name || '').toLowerCase();
      const url = (n.url || '').toLowerCase();
      if (nm.includes(kw) || url.includes(kw)) results.push(n);
    }
    for (const child of n.children || []) search(child);
  }
  search(node);
  return results;
}

function getMaxId(node) {
  let max = parseInt(node.id || '0', 10);
  for (const child of node.children || []) {
    max = Math.max(max, getMaxId(child));
  }
  return max;
}

function remapIds(node, idCounter) {
  const copy = JSON.parse(JSON.stringify(node));
  copy.id = String(idCounter.value++);
  if (copy.guid) copy.guid = crypto.randomUUID();

  const now = String(BigInt(Date.now()) * 1000n + 11644473600000000000n);
  if (copy.type === 'folder') {
    copy.date_added = now;
    copy.date_modified = now;
  } else {
    copy.date_added = now;
  }
  delete copy.meta_info;

  if (copy.children) {
    copy.children = copy.children.map(c => remapIds(c, idCounter));
  }
  return copy;
}

function calculateChecksum(data) {
  const content = JSON.stringify(data.roots, (k, v) => {
    // 保持和 Python 一致：按 key 排序
    if (v && typeof v === 'object' && !Array.isArray(v)) {
      return Object.keys(v).sort().reduce((acc, key) => {
        acc[key] = v[key];
        return acc;
      }, {});
    }
    return v;
  });
  return crypto.createHash('md5').update(content, 'utf-8').digest('hex');
}

function listAllFolders(node, folderPath = '') {
  const folders = [];
  if (node.type === 'folder') {
    const fullPath = folderPath ? `${folderPath}/${node.name}` : node.name;
    const bookmarkCount = (node.children || []).filter(c => c.type === 'url').length;
    folders.push({ path: fullPath, count: bookmarkCount });
    for (const child of node.children || []) {
      folders.push(...listAllFolders(child, fullPath));
    }
  }
  return folders;
}

// ── 列出收藏夹 ──────────────────────────────────────────────
function listFolders(sourceBrowser) {
  const sourcePath = getBookmarksPath(sourceBrowser);
  if (!sourcePath) return false;

  const data = loadChromiumBookmarks(sourcePath);
  console.log(`\n📂 ${sourceBrowser.toUpperCase()} 的所有收藏夹:\n`);

  const allFolders = [];
  for (const [rootName, rootNode] of Object.entries(data.roots || {})) {
    if (rootNode && typeof rootNode === 'object') {
      allFolders.push(...listAllFolders(rootNode));
    }
  }

  if (!allFolders.length) {
    console.log('  (没有收藏夹)');
    return true;
  }

  for (const { path: fp, count } of allFolders.sort((a, b) => a.path.localeCompare(b.path))) {
    const depth = fp.split('/').length - 1;
    const indent = '  '.repeat(depth);
    const name = fp.split('/').pop();
    console.log(`${indent}📁 ${name} (${count} 个书签)`);
  }

  console.log(`\n总计: ${allFolders.length} 个收藏夹`);
  return true;
}

// ── 主迁移逻辑 ──────────────────────────────────────────────
function migrateBookmarks({ source, target, folder, keyword, dryRun }) {
  const sourcePath = getBookmarksPath(source);
  const targetPath = getBookmarksPath(target);

  if (!sourcePath || !targetPath) return false;

  console.log(`📂 源: ${sourcePath}`);
  console.log(`📂 目标: ${targetPath}`);

  // 加载源书签
  let sourceData;
  if (source === 'firefox') {
    loadFirefoxBookmarks(sourcePath);
    return false;
  } else if (source === 'safari') {
    loadSafariBookmarks(sourcePath);
    return false;
  } else {
    sourceData = loadChromiumBookmarks(sourcePath);
  }

  // 加载目标书签
  if (target === 'firefox' || target === 'safari') {
    console.error('⚠️ 目标浏览器为 Firefox/Safari，暂不支持写入');
    return false;
  }
  const targetData = loadChromiumBookmarks(targetPath);

  // 查找要迁移的内容
  let itemsToMigrate = [];

  if (folder) {
    let found = null;
    for (const rootNode of Object.values(sourceData.roots || {})) {
      if (rootNode && typeof rootNode === 'object') {
        found = findFolder(rootNode, folder);
        if (found) break;
      }
    }
    if (!found) {
      console.error(`❌ 未找到文件夹 '${folder}'`);
      console.error('\n可用文件夹:');
      for (const rootNode of Object.values(sourceData.roots || {})) {
        if (rootNode && typeof rootNode === 'object') {
          for (const { path: fp, count } of listAllFolders(rootNode)) {
            console.error(`  📁 ${fp} (${count} 个书签)`);
          }
        }
      }
      return false;
    }
    itemsToMigrate = [found];
    console.log(`✅ 找到文件夹 '${folder}'，包含 ${found.children?.length || 0} 个子项`);
  } else if (keyword) {
    for (const rootNode of Object.values(sourceData.roots || {})) {
      if (rootNode && typeof rootNode === 'object') {
        itemsToMigrate.push(...findBookmarksByKeyword(rootNode, keyword));
      }
    }
    if (!itemsToMigrate.length) {
      console.error(`❌ 未找到包含关键词 '${keyword}' 的书签`);
      return false;
    }
    console.log(`✅ 找到 ${itemsToMigrate.length} 个匹配的书签`);
  } else {
    // 全量迁移
    for (const rootNode of Object.values(sourceData.roots || {})) {
      if (rootNode && typeof rootNode === 'object' && rootNode.children) {
        itemsToMigrate.push(...rootNode.children);
      }
    }
    console.log(`✅ 全量迁移，共 ${itemsToMigrate.length} 个顶级项`);
  }

  // 预览模式
  if (dryRun) {
    console.log('\n🔍 预览模式（不执行写入）:');
    for (const item of itemsToMigrate) {
      if (item.type === 'folder') {
        console.log(`  📁 ${item.name} (${item.children?.length || 0} 个子项)`);
      } else {
        console.log(`  🔗 ${item.name || 'Unnamed'} - ${item.url || 'No URL'}`);
      }
    }
    return true;
  }

  // 备份
  const timestamp = new Date().toISOString().replace(/[-:T]/g, '').slice(0, 14);
  const backupPath = `${targetPath}.bak_${timestamp}`;
  fs.copyFileSync(targetPath, backupPath);
  console.log(`\n💾 已备份目标书签到: ${backupPath}`);

  // 执行迁移
  let maxId = 0;
  for (const rootNode of Object.values(targetData.roots || {})) {
    if (rootNode && typeof rootNode === 'object') {
      maxId = Math.max(maxId, getMaxId(rootNode));
    }
  }

  const idCounter = { value: maxId + 1 };
  const bookmarkBar = targetData.roots?.bookmark_bar;
  if (!bookmarkBar) {
    console.error('❌ 目标书签文件格式异常：缺少 bookmark_bar');
    return false;
  }
  if (!bookmarkBar.children) bookmarkBar.children = [];

  for (const item of itemsToMigrate) {
    bookmarkBar.children.push(remapIds(item, idCounter));
  }

  // 更新 checksum 并写入
  targetData.checksum = calculateChecksum(targetData);
  fs.writeFileSync(targetPath, JSON.stringify(targetData, null, 3), 'utf-8');

  console.log(`✅ 迁移完成！已写入 ${itemsToMigrate.length} 个项目到 ${target}`);
  console.log(`📝 请重启 ${target} 以刷新书签`);
  return true;
}

// ── CLI ─────────────────────────────────────────────────────
function showHelp() {
  console.log(`
浏览器收藏夹迁移工具 (Node.js)

用法:
  node migrate-bookmarks.mjs [选项]

选项:
  -s, --source  <浏览器>   源浏览器 (chrome/edge/tabbit/arc/brave)
  -t, --target  <浏览器>   目标浏览器 (chrome/edge/tabbit/arc/brave)
  -f, --folder  <名称>     指定要迁移的文件夹名称
  -k, --keyword <关键词>   按关键词过滤书签
  -d, --dry-run            预览模式，不实际写入
  -l, --list               列出源浏览器的所有文件夹
  -h, --help               显示帮助

示例:
  # 列出 Chrome 所有收藏夹
  node migrate-bookmarks.mjs -s chrome -l

  # 预览迁移
  node migrate-bookmarks.mjs -s chrome -t tabbit -f "Frontend-GItHubBlog" -d

  # 执行迁移
  node migrate-bookmarks.mjs -s chrome -t tabbit -f "Frontend-GItHubBlog"

  # 全量迁移
  node migrate-bookmarks.mjs -s chrome -t edge

  # 按关键词迁移
  node migrate-bookmarks.mjs -s chrome -t arc -k github
`);
}

function parseArgs(argv) {
  const args = {};
  for (let i = 2; i < argv.length; i++) {
    const arg = argv[i];
    switch (arg) {
      case '-s': case '--source':  args.source  = argv[++i]; break;
      case '-t': case '--target':  args.target  = argv[++i]; break;
      case '-f': case '--folder':  args.folder  = argv[++i]; break;
      case '-k': case '--keyword': args.keyword = argv[++i]; break;
      case '-d': case '--dry-run': args.dryRun  = true; break;
      case '-l': case '--list':    args.list    = true; break;
      case '-h': case '--help':    args.help    = true; break;
      default:
        if (arg.startsWith('-')) {
          console.error(`❌ 未知选项: ${arg}`);
          process.exit(1);
        }
    }
  }
  return args;
}

function main() {
  const args = parseArgs(process.argv);

  if (args.help || process.argv.length <= 2) {
    showHelp();
    process.exit(0);
  }

  if (args.list) {
    if (!args.source) {
      console.error('❌ 使用 --list 时需要指定 --source');
      process.exit(1);
    }
    const ok = listFolders(args.source);
    process.exit(ok ? 0 : 1);
  }

  if (!args.source || !args.target) {
    console.error('❌ 必须指定 --source 和 --target');
    showHelp();
    process.exit(1);
  }

  const ok = migrateBookmarks(args);
  process.exit(ok ? 0 : 1);
}

main();
